### PR TITLE
feat: Improvements to GitHub Actions workflow

### LINKED ISSUE #1144: PR Checking with GitHub Actions
## Acceptance Criteria 

- [x] should create clusters on GKE
- [x] github actions workflow that runs smoke, CATS, SITS and brain tests for PRs
- [x] runs diego
- [x] runs eirini
- [ ] doesn't run for PRs from a fork
- [x] should cleanup any cloud resources
- [x] should not be mandatory for merges

Use the work done here as a starting point: https://github.com/mook-as/kubecf/actions/

> [x] GKE clusters should have only one node

(Edited by @mook-as: Making checkboxes)

### PR DESCRIPTION
**Pull request by mook-as**
## Description
~This pull request will change our GitHub Actions workflow to include at least the basic tests, to the point where we might start deprecating our Concourse pipelines.~

~For now, though, I just have a design of how we might handle the build triggering and cluster allocation; I'd like some feedback before actually implementing things.   Hence, this is a draft with just markdown changes.  I'll edit the description once we agree and I've implemented it reasonably.~

I am now rescoping this to _just_ the changes described to scheduling the runs; this is already enough of a change that it's a good chunk to review.  Also, I can work on the tests independently of this PR (in my fork, manually triggering runs).

## Motivation and Context
Part of #1144 (but will no longer fix it by itself).

## How Has This Been Tested?

I have been running this in my fork; [sample run](https://github.com/mook-as/kubecf/actions/runs/338099108).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code has security implications.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

This has security implications in the sense of this may trigger runs on CI of potentially untrusted code.  So it's security of our infrastructure (and from there our code), not of the product (unless they manage to change our code).

### CODE REVIEW

**Comment by mook-as** [comment]
FWIW, I would probably also move the linting to its own file, since that doesn't need to allocate a cluster; this would mean we get quicker feedback on linting errors, but still run the full tests once we have a cluster.  That will need to also be after PR approval, though.

The `pull-request-schedule.yaml` step will probably need to written as a script rather than actions YAML, since it will need to be able to make GitHub API calls — as far as I can tell, sorting issue by last-label-apply-time doesn't really have a good API.

**Comment by jandubois** [comment]
> That will need to also be after PR approval, though.

Why can't we run just linting before PR approval? At least then we won't start reviewing until all linting errors are fixed. 😄

**Comment by mook-as** [comment]
> > That will need to also be after PR approval, though.
> 
> Why can't we run just linting before PR approval? At least then we won't start reviewing until all linting errors are fixed. 😄

Because the linting script would come from the PR, and it could touch `GITHUB_TOKEN`, like everything else, right? Hmm. At least, I assumed the token would be available; now that I think about it though, it might be fine (because the workflow definition is still from `master`, and I don't see us exposing that for any reason…)

**Comment by viccuad** [comment]
The build triggering and cluster automation sounds perfect, gives us the blocking pool that we never implemented in Concourse.

File: .github/workflows/pull-request-schedule.yaml
Comment (viccuad) [inline .github/workflows/pull-request-schedule.yaml:39]: Does this respect newline characters correctly?

File: .github/workflows/pull-request-lint.yaml
Comment (viccuad) [inline .github/workflows/pull-request-lint.yaml:5]: One could still attack some vulnerability of the linters themselves, to run arbitrary code on PR push. But I suppose that's a small enough attack surface.

File: .github/workflows/pull-request-schedule/trigger-builds.js
Comment (viccuad) [inline .github/workflows/pull-request-schedule/trigger-builds.js:10]: Does this `last: 10` in the query for OPEN pullRequests means to obtain only the last 10 PRs? Or is it a pagination thing or the like?

**Review by viccuad — DISMISSED** [review]
Looks good here! approving, but note that I just learnt how to use map, reduce, filter, etc on js. Cool API polling and queue implementation with the GH labels, states and age. Yet, I hope we don't grow the amount of Javascript :sweat_smile: .

Things that are outside the scope of this PR:

- In pull-request-ci.yaml, I think there may be misconfigurations because we are not setting `BACKEND` or `CLUSTER_NAME`. If they aren't set, it will go with the default  `BACKEND=kind`, which will set `CLUSTER_NAME=kind` and create a _buildkind_ . I doubt we want this because when we create the cluster we need to set `BACKEND=gke` which will give us a _buildgke_ to operate on the rest of targets. Maybe we could add them to the global `env` in pull-request-ci.yaml?

- The strategy matrix perhaps needs to be expanded and maybe reworked into 2 lanes (diego & eirini, as diego has testsuites that only apply to it) and then add an experimental flag for things that can continue on error. Without it, we may fail in brains and sits.

File: .github/workflows/pull-request-schedule.yaml
Comment (mook-as) [reply .github/workflows/pull-request-schedule.yaml:39]: There are no new line characters involved: `GKE_CRED_JSON` (as exported to `GITHUB_ENV`) is the path to the temporary file we created.  (There's a new line created by `echo` but I don't think that's not what you meant.)

File: .github/workflows/pull-request-schedule/trigger-builds.js
Comment (mook-as) [reply .github/workflows/pull-request-schedule/trigger-builds.js:10]: You're right, this means we only get 10 PRs (with the label); that's obviously wrong (I forgot to change the testing values I used).  I should double check what order this is in (by testing with a different label), and see if picking the _first_ PRs would make sense; since we will eventually remove the labels (when the PRs get scheduled) that should work fine.

File: .github/workflows/pull-request-lint.yaml
Comment (mook-as) [reply .github/workflows/pull-request-lint.yaml:5]: Yes, but the linter runs don't get to have any secrets (since we don't need to create clusters to test them, and we check out the source with `persist-credentials: false`).  The issue with the workflow file is that _it_ can have access to secrets (e.g. `github.token`), which it may choose to export to scripts.

…I think; it'll probably be a good idea to double check the environment in that case.  I'll try to do that.  Either way it sounds like I need to expand the comment to say ↑.

File: .github/workflows/pull-request-schedule/trigger-builds.js
Comment (mook-as) [reply .github/workflows/pull-request-schedule/trigger-builds.js]: Fixed in a40333443489f8407f4b019b8e1b25669dcc0250

File: .github/workflows/pull-request-lint.yaml
Comment (mook-as) [reply .github/workflows/pull-request-lint.yaml]: Added more comments in fcb37225b6f4f4e2c9dc674b038ddf60abbd3580

File: .github/workflows/pull-request-schedule.yaml
Comment (viccuad) [reply .github/workflows/pull-request-schedule.yaml:39]: doh true, it's a file. That answers my question.

**Review by viccuad — APPROVED** [review]

**Pull request merged**
